File: android/Streamo/app/src/main/java/com/streamo/app/provider/IntroSkipClient.kt
package com.streamo.app.provider

import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Fetches skip-intro / skip-credits timestamps from TheIntroDB (v3 API).
 *
 * Keyed by TMDB id — exactly what the player already carries, so no id
 * conversion is needed. Reads are keyless (an API key only raises the rate
 * limit); a missing/failed lookup returns null so the player simply shows no
 * skip button and nothing regresses.
 *
 * Offline playback is intentionally out of scope: the caller (PlayerViewModel)
 * only queries this client for online TMDB sources.
 *
 * Android port of iOS IntroSkipClient.swift.
 */
@Singleton
class IntroSkipClient @Inject constructor(
    private val client: OkHttpClient
) {

    /** Segment boundaries in milliseconds. */
    data class Segments(
        val introStartMs: Long? = null,
        val introEndMs: Long? = null,
        val creditsStartMs: Long? = null
    ) {
        val isEmpty: Boolean get() = introEndMs == null && creditsStartMs == null
    }

    private val gson = Gson()

    /** Look up segments for a movie or TV episode. */
    suspend fun fetch(
        tmdbId: Int,
        isMovie: Boolean,
        season: Int,
        episode: Int,
        durationMs: Long?
    ): Segments? = withContext(Dispatchers.IO) {
        if (tmdbId <= 0) return@withContext null

        val urlBuilder = StringBuilder("https://api.theintrodb.org/v3/media?tmdb_id=$tmdbId")
        if (!isMovie) {
            urlBuilder.append("&season=$season&episode=$episode")
        }
        if (durationMs != null && durationMs > 0) {
            urlBuilder.append("&duration_ms=$durationMs")
        }

        val request = Request.Builder()
            .url(urlBuilder.toString())
            .header("User-Agent", "Streamo/1.0")
            .header("Accept", "application/json")
            .build()

        val response = try {
            client.newCall(request).apply {
                timeout().timeout(8, TimeUnit.SECONDS)
            }.execute()
        } catch (_: Exception) {
            return@withContext null
        }

        if (!response.isSuccessful) {
            response.close()
            return@withContext null
        }

        val body = try {
            response.body?.string()
        } catch (_: Exception) {
            null
        } finally {
            response.close()
        }

        if (body.isNullOrBlank()) return@withContext null

        val decoded = try {
            gson.fromJson(body, MediaResponse::class.java)
        } catch (_: Exception) {
            null
        } ?: return@withContext null

        val segments = Segments(
            introStartMs = decoded.intro?.firstOrNull()?.startMs,
            introEndMs = decoded.intro?.firstOrNull()?.endMs,
            creditsStartMs = decoded.credits?.firstOrNull()?.startMs
        )
        if (segments.isEmpty) null else segments
    }

    private data class MediaResponse(
        val intro: List<Segment>?,
        val credits: List<Segment>?
    ) {
        data class Segment(
            @SerializedName("start_ms") val startMs: Long?,
            @SerializedName("end_ms") val endMs: Long?
        )
    }
}

File: ios/Streamo/Streamo/Provider/IntroSkipClient.swift
import Foundation

/// Fetches skip-intro / skip-credits timestamps from TheIntroDB (v3 API).
///
/// Keyed by TMDB id — exactly what `PlaybackRequest` already carries, so no id
/// conversion is needed. Reads are keyless (an API key only raises the rate
/// limit); a missing/failed lookup returns `nil` so the player simply shows no
/// skip button and nothing regresses.
///
/// Anime is intentionally out of scope here: for `.animeUnity` the request's
/// `tmdbId` actually holds the AnimeUnity entry id, not a real TMDB id, so the
/// caller (`PlaybackController`) only queries this client for `.tmdb` sources.
actor IntroSkipClient {
    static let shared = IntroSkipClient()

    /// Segment boundaries in **seconds**. `introStart` is nil when the intro
    /// runs from the very start (movies often report this); `creditsStart` is
    /// nil when there are no credits data.
    struct Segments: Sendable, Equatable {
        var introStart: Double?
        var introEnd: Double?
        var creditsStart: Double?

        /// True when there's nothing actionable — lets the caller drop it.
        var isEmpty: Bool { introEnd == nil && creditsStart == nil }
    }

    private let session: URLSession
    private let requestTimeout: TimeInterval = 8

    init(session: URLSession = .shared) {
        self.session = session
    }

    /// Look up segments for a movie or TV episode. `durationMs` is optional but
    /// recommended: TheIntroDB uses it to pick the matching release cut when a
    /// title has multiple versions.
    func fetch(tmdbId: Int, isMovie: Bool, season: Int, episode: Int, durationMs: Double?) async -> Segments? {
        guard tmdbId > 0 else { return nil }
        var comps = URLComponents(string: "https://api.theintrodb.org/v3/media")!
        var items = [URLQueryItem(name: "tmdb_id", value: String(tmdbId))]
        if !isMovie {
            items.append(URLQueryItem(name: "season", value: String(season)))
            items.append(URLQueryItem(name: "episode", value: String(episode)))
        }
        if let ms = durationMs, ms > 0 {
            items.append(URLQueryItem(name: "duration_ms", value: String(Int(ms))))
        }
        comps.queryItems = items
        guard let url = comps.url else { return nil }

        var request = URLRequest(url: url)
        request.timeoutInterval = requestTimeout
        request.setValue("Streamo/1.0", forHTTPHeaderField: "User-Agent")   // API rejects empty UA
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        guard let (data, response) = try? await session.data(for: request),
              let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
              let decoded = try? JSONDecoder().decode(MediaResponse.self, from: data) else {
            return nil
        }

        func seconds(_ ms: Double?) -> Double? { ms.map { $0 / 1000 } }
        let segments = Segments(
            introStart: seconds(decoded.intro?.first?.startMs),
            introEnd: seconds(decoded.intro?.first?.endMs),
            creditsStart: seconds(decoded.credits?.first?.startMs)
        )
        return segments.isEmpty ? nil : segments
    }

    // MARK: - Wire format

    private struct MediaResponse: Decodable {
        let intro: [Segment]?
        let credits: [Segment]?

        struct Segment: Decodable {
            let startMs: Double?
            let endMs: Double?
            enum CodingKeys: String, CodingKey {
                case startMs = "start_ms"
                case endMs = "end_ms"
            }
        }
    }
}
